### PR TITLE
Runtime config in Docker

### PR DESCRIPTION
**Pull request by constantinius**
This PR adds a script to be run by the nginx entrypoint. This writes seval env variables to the config.js.

If built with the `DYNAMIC_CONFIG=true` option, then the config.js will be loaded on start of the app.

Based on #346 and #347

### CODE REVIEW

**Comment by constantinius** [comment]
The issue is that we don't have node anymore in the final image. We just have nginx + shell, node is just available for the build-stage.

There is an additional effort for maintenance, that is true.

I tried to make some shell magic to make it easier to convert from shell variable style to expected JS/JSON in the config.js. But we could also get rid of that, loop over all env variables with `SB_` prefix and simply write the values to `config.js` without any conversion. But then the env vars must be of exactly pre-formatted.

To be honest, I prefer the way it is currently done in this PR. The UPPERCASE variables signify that they are not the same as the JS values and that there is a mapping step.

**Review by m-mohr — CHANGES_REQUESTED** [review]
I don't want to maintain this. An approach where just the defined ENV variables (starting with SB_) are written to the config would also be less error prone as it only affects what had been specified. I also don't want to document yet another way of variables. We already have so many ways, so I'd really prefer to align with what we currently have.

**Comment by waystilos** [comment]
If I'm understanding the configuration correctly, it seems that any changes to the variable mappings or logic would require manual modification of the script. While I can see how this approach adds flexibility, I'm not sure if it's worth the maintenance. 

Maintaining the script for future changes to the configuration options might become challenging, especially as the project evolves and it snowballs. Happy discuss further, especially if I'm not understanding the benefits.

**Comment by constantinius** [comment]
I merged #358 into this, now all `SB_` prefixed env variables are written to the config, with interpolations applied via the parameter schema.

@m-mohr does it make more sense now?

**Comment by m-mohr** [comment]
@constantinius Yeah, makes more sense although I don't understand everything that's going on.

Could you please rebase? Then I'd leave the final review up to @waystilos and @geospatial-jeff.

**Comment by m-mohr** [comment]
This is the last remaining open PR/issue for 3.1.0. I hope we can still get this in before end of August.

File: docker-entrypoint.sh
Comment (waystilos) [inline docker-entrypoint.sh:1]: Purely a suggestion for error handling here:

# echo a string, handling different types
safe_echo() {
    local value="$1"
    if [ -z "$value" ]; then
        echo -n "null"
    elif printf '%s\n' "$value" | grep -qE '\n.+\n$'; then
        echo -n "\`$value\`"
    else
        echo -n "'$value'"
    fi
}

#  handle boolean
bool() {
    case "$1" in
        true | TRUE | yes | t | True)
            echo -n true ;;
        false | FALSE | no | n | False)
            echo -n false ;;
        *)
            echo "Err: Unknown boolean value \"$1\"" >&2
            exit 1 ;;
    esac
}

# handle array values
array() {
    local value="$1"
    local arraytype="$2"
    if [ -z "$value" ]; then
        echo -n "[]"
    else
        case "$arraytype" in
            string)
                echo -n "['$(echo "$value" | sed "s/,/', '/g")']"
                ;;
            *)
                echo -n "[$value]"
                ;;
        esac
    fi
}

# handle object values
object() {
    local value="$1"
    if [ -z "$value" ]; then
        echo -n "null"
    else
        echo -n "$value"
    fi
}

config_schema=$(cat /etc/nginx/conf.d/config.schema.json)

# Iterate over environment variables with "SB_" prefix
env -0 | cut -f1 -d= | tr '\0' '\n' | grep "^SB_" | {
    echo "window.STAC_BROWSER_CONFIG = {"
    while IFS='=' read -r name; do
        # Strip the prefix
        argname="${name#SB_}"
        # Read the variable's value
        value="$(eval "echo \"\$$name\"")"

        # Get the argument type from the schema
        argtype="$(echo "$config_schema" | jq -r ".properties.$argname.type[0]")"
        arraytype="$(echo "$config_schema" | jq -r ".properties.$argname.items.type[0]")"

        # Encode key/value
        echo -n "  $argname: "
        case "$argtype" in
            string)
                safe_echo "$value"
                ;;
            boolean)
                bool "$value"
                ;;
            integer | number)
                object "$value"
                ;;
            array)
                array "$value" "$arraytype"
                ;;
            *)
                safe_echo "$value"
                ;;
        esac
        echo ","
    done
    echo "}"
} > /usr/share/nginx/html/config.js

**Review by waystilos — APPROVED** [review]
lgtm.

**Review by m-mohr — APPROVED** [review]

**Pull request merged**